### PR TITLE
RoutePattern everywhere

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

File: samples/DispatcherSample.Web/Startup.cs [hunk 1]
```diff
@@ -39,10 +40,8 @@ public void ConfigureServices(IServiceCollection services)
                             response.ContentLength = payloadLength;
                             return response.Body.WriteAsync(_homePayload, 0, payloadLength);
                         },
-                        "/",
+                        RoutePatternFactory.Parse("/"),
                         new RouteValueDictionary(),
-                        new RouteValueDictionary(),
-                        new List<MatchProcessorReference>(),
```
Comment (rynowak) [inline samples/DispatcherSample.Web/Startup.cs]: I'm open to discussion about what constructors we should have here. I'm also sensitive to the fact that we have like 9 parameters. We should try to make a decision here that will carry us through preview1.

File: src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs [hunk 2]
```diff
@@ -20,145 +20,120 @@ internal class DefaultMatchProcessorFactory : MatchProcessorFactory
...
-            // ConstraintArgument: \d+
-
-            (var constraintName, var constraintArgument) = Parse(matchProcessorReference.ConstraintText);
+            // value: regex(\d+)
+            // name: regex
+            // argument: \d+
+            (var name, var argument) = Parse(value);
 
-            if (!_options.ConstraintMap.TryGetValue(constraintName, out var constraintType))
+            if (!_options.ConstraintMap.TryGetValue(name, out var type))
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs]: type -> constraintType

File: src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs [hunk 3]
```diff
@@ -20,145 +20,120 @@ internal class DefaultMatchProcessorFactory : MatchProcessorFactory
...
-                    $"An error occurred while trying to create an instance of route constraint '{constraintType.FullName}'.",
-                    exception);
-            }
+            var message = Resources.FormatRoutePattern_InvalidStringConstraintReference(
+                type,
+                name,
+                typeof(IRouteConstraint),
+                typeof(MatchProcessor));
+            throw new InvalidOperationException(message);
+
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs]: nit: remove

File: src/Microsoft.AspNetCore.Routing/Matchers/RouteConstraintMatchProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        public RouteConstraintMatchProcessor(string parameterName, IRouteConstraint constraint)
+        {
+            ParameterName = parameterName;
+            Constraint = constraint;
+        }
+
+        public string ParameterName { get; }
+
+        public IRouteConstraint Constraint { get; }
+
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing/Matchers/RouteConstraintMatchProcessor.cs]: nit: too liney

File: src/Microsoft.AspNetCore.Routing/Matchers/RouteConstraintMatchProcessor.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing/Matchers/RouteConstraintMatchProcessor.cs]: nit: too liney

File: src/Microsoft.AspNetCore.Routing/Matchers/RouteConstraintMatchProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+using System;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Routing.Matchers
+{
+    internal class RouteConstraintMatchProcessor: MatchProcessor
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing/Matchers/RouteConstraintMatchProcessor.cs]: nit: space

File: src/Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs [hunk 7]
```diff
@@ -92,6 +129,37 @@ private static int ComputeOutboundPrecedenceDigit(TemplateSegment segment)
...
+        {
+            if (pathSegment.Parts.Count > 1)
+            {
+                return 4;
+            }
+
+            var part = pathSegment.Parts[0];
+            if (part.IsLiteral)
+            {
+                return 5;
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs]: why 5 (and why 4 above it)

File: src/Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs [hunk 8]
```diff
@@ -92,6 +129,37 @@ private static int ComputeOutboundPrecedenceDigit(TemplateSegment segment)
...
+            }
+            else if (part is RoutePatternParameterPart parameterPart)
+            {
+                Debug.Assert(parameterPart != null);
+                var digit = parameterPart.IsCatchAll ? 1 : 3;
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs]: why these numbers too

File: test/Microsoft.AspNetCore.Routing.Tests/Matchers/RouteConstraintMatchProcessorTest.cs [hunk 9]
```diff
@@ -0,0 +1,9 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Routing.Matchers
+{
+    public class RouteConstraintMatchProcessorTest
```
Comment (JamesNK) [inline test/Microsoft.AspNetCore.Routing.Tests/Matchers/RouteConstraintMatchProcessorTest.cs]: hmm?

File: src/Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs (hunk 7 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs]: See comments elsewhere in this file.

File: src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs]: It's pretty clear right? What other type would it get confused with?

File: src/Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs (hunk 8 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs]: See comments elsewhere in this file

File: src/Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs (hunk 7 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs]: Sorry, I should have explained this. None of what's here is new code. It's existing logic that can't be super-re-used without creating a nother overload.

File: test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs [hunk 10]
```diff
@@ -785,50 +748,6 @@ public void GetLink_OptionalParameter_ParameterNotPresentInValues()
...
-                new[] { endpoint },
-                context.ExplicitValues,
-                context.AmbientValues);
-
-            // Assert
-            Assert.Equal("/Home/Index/products", link);
-        }
-
-        [Fact]
-        public void GetLink_OptionalParameter_ParameterNotPresentInValues_PresentInDefaults()
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs]: These two tests are not valid, and the fact that there were possible to write is a bug that I've fixed in RoutePattern. You can't have a default value for an optional parameter.

File: test/Microsoft.AspNetCore.Routing.Tests/Matchers/RouteConstraintMatchProcessorTest.cs (hunk 9 above)
Comment (rynowak) [reply test/Microsoft.AspNetCore.Routing.Tests/Matchers/RouteConstraintMatchProcessorTest.cs]: Oh oops

File: src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs (hunk 2 above)
Comment (JamesNK) [reply src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs]: sure. probably should have `nit:`ed it 😄

File: samples/DispatcherSample.Web/Startup.cs (hunk 1 above)
Comment (JamesNK) [reply samples/DispatcherSample.Web/Startup.cs]: Will your average user ever want to create an endpoint themselves? If so then maybe one or two shorter ones? I don't have any strong opinions.

File: src/Microsoft.AspNetCore.Routing/Matchers/RouteConstraintMatchProcessor.cs (hunk 4 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing/Matchers/RouteConstraintMatchProcessor.cs]: Fixed but not showing in the diff

File: samples/DispatcherSample.Web/Startup.cs (hunk 1 above)
Comment (rynowak) [reply samples/DispatcherSample.Web/Startup.cs]: I don't think that's likely in 2.2 the audience for now is frameworks/middlewarelikes.

**Comment by rynowak** [comment]
Updated. I added a few tersts for the RCMP. They are the gross Moq style, and I hate that but I wanted there to be **something** that unit tests that class.

File: src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs [hunk 11]
```diff
@@ -20,145 +20,119 @@ internal class DefaultMatchProcessorFactory : MatchProcessorFactory
...
             string parameterName,
             bool optional,
-            Type constraintType,
-            string constraintArgument)
+            IRouteConstraint constraint,
+            string argument)
         {
-            var routeConstraint = DefaultInlineConstraintResolver.CreateConstraint(constraintType, constraintArgument);
-            var matchProcessor = new MatchProcessorReference(parameterName, routeConstraint).MatchProcessor;
+            var matchProcessor = (MatchProcessor)new RouteConstraintMatchProcessor(parameterName, constraint);
```
Comment (kichalla) [inline src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs]: After this, call into the other InitializeMatchProcessor method

File: src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs [hunk 12]
```diff
@@ -20,145 +20,119 @@ internal class DefaultMatchProcessorFactory : MatchProcessorFactory
...
         {
-            if (matchProcessorReference == null)
+            if (value == null)
             {
-                throw new ArgumentNullException(nameof(matchProcessorReference));
+                throw new ArgumentNullException(nameof(value));
             }
 
-            if (matchProcessorReference.MatchProcessor != null)
+            return InitializeMatchProcessor(parameterName, optional, value, null);
```
Comment (kichalla) [inline src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs]: nit: named argument

**Pull request merged**